marker_rustc_driver/src/main.rs

```diff
@@ -71,6 +95,49 @@ impl rustc_driver::Callbacks for MarkerCallback {
     }
 }
 
+fn register_tracked_env(sess: &mut rustc_session::parse::ParseSess, vars: &[(&'static str, String)]) {
+    use rustc_span::Symbol;
+    let env = sess.env_depinfo.get_mut();
+
+    for (key, value) in vars {
+        env.insert((Symbol::intern(key), Some(Symbol::intern(value))));
+    }
+}
+
+fn register_tracked_files(sess: &mut rustc_session::parse::ParseSess, lint_crates: &[LintCrateInfo]) {
+    use rustc_span::Symbol;
+
+    let files = sess.file_depinfo.get_mut();
+
+    // Cargo sets the current directory, to the folder containing the `Cargo.toml`
+    // file of the compiled crate. Therefore, we can use the relative path.
+    let cargo_file = Path::new("./Cargo.toml");
+    if cargo_file.exists() {
+        files.insert(Symbol::intern("./Cargo.toml"));
+    }
+
+    // At this point, the lint crate paths should be absolute. Therefore, we
+    // can just throw it in the `files` set.
+    for lint_crate in lint_crates {
+        if let Some(name) = lint_crate.path.to_str() {
+            files.insert(Symbol::intern(name));
```

Comment:
I wonder if it track the directory recursively. IIRC there was some ~~bug~~ feature in cargo that if your use `rerun-if-changed` on a directory in a build script, then only the dir metadata or just the first level of entries will be tracked.

Reply:
I've recently talked to someone from a project that always uses the rustc/clippy driver directly. For Marker they would do the same. In such a case, a lint crate might be specified that is not inside the lints directory of the project.

Doing it this way doesn't put any restrictions on the user for little to no cost to us :)